numpy/ma/core.py

```diff
@@ -3305,6 +3305,8 @@ def _scalar_heuristic(arr, elem):
                     # which should have shape () to avoid breaking several
                     # methods. There is no great way out, so set to
                     # first element.  See issue #6723.
+                    # something like gh-15895 has happened if this fails
+                    assert isinstance(dout._fill_value, np.ndarray)
```

Comment:
This seems wrong. If there is a way for `dout._fill_value` to ever not be an `ndarray`, then this should raise a `ValueError`, not an `AssertionError`. And it would be even better if there was a test to show how this could ever happen.

Reply:
> it would be even better if there was a test to show how this could ever happen.

It can't ever happen, as far as a I know - but this code is bug-prone, and this makes it easier for us to work out the issue if it does fire.

We could use `RuntimeError("internal numpy error")` or similar if you prefer.

Reply:
+1. As far as I can tell we have no other asserts in the code, we do have RuntimeErrors for this kind of thing.